xapian-core/include/xapian/cluster.h

```diff
@@ -0,0 +1,489 @@
+    friend class ClusterSet;
+  public:
+    ClusterSet cset;
+    Xapian::doccount index;
+
+    ClusterSetIterator(ClusterSet &cset_, Xapian::doccount index_)
+	: cset(cset_), index(index_) {}
+
+    clusterid operator*() const;
+
+    // Returns the id of the current cluster
+    clusterid get_clusterid() {
+	return index;
+    }
+
+    Cluster get_cluster();
+
+    // Prefix increment operator for this iterator
+    ClusterSetIterator & operator++() {
+	index++;
+	return *this;
+    }
+
+    // Postfix increment operator for this iterator
+    ClusterSetIterator operator++(int) {
+	ClusterSetIterator *c = this;
+	index++;
+	return *c;
+    }
+
+    // Prefix decrement operator for this iterator
+    ClusterSetIterator & operator--() {
+	index--;
+	return *this;
+    }
+
+    // Postfix decrement operator for this iterator
+    ClusterSetIterator operator--(int) {
+	ClusterSetIterator c = *this;
+	index--;
+	return c;
+    }
+};
+
+// Equality test for ClusterSetIterator
+inline bool operator==(const ClusterSetIterator &a, const ClusterSetIterator &b) {
+    return (a.index == b.index);
+}
+
+// Inequality test for ClusterSetIterator
+inline bool operator!=(const ClusterSetIterator &a, const ClusterSetIterator &b) {
+    return !(a == b);
+}
+
+// Equality test for DocumentSetIterator
+inline bool operator==(const DocumentSetIterator &a, const DocumentSetIterator &b) {
+    return (a.index == b.index);
+}
+
+// Inequality test for DocumentSetIterator
+inline bool operator!=(const DocumentSetIterator &a, const DocumentSetIterator &b) {
+    return !(a == b);
+}
+
+/** Round Robin clusterer:
+ *  This clusterer is a minimal clusterer which will cluster documents as -
+ *  ith document goes to the (i%N)th cluster where N is the number of clusters and
+ *  0 <= i <= k; where k is the number of documents
+ */
+class XAPIAN_VISIBILITY_DEFAULT RoundRobin {
+  public:
+    ~RoundRobin();
+
+    // Implements RoundRobin clustering
+    ClusterSet cluster(MSet &mset, unsigned int num_of_clusters);
+
+    // Returns the description of the clusterer
+    std::string get_description();
+};
+
+/** Kmeans clusterer:
+ *  This clusterer implements the K-Means clustering algorithm
+ */
+class XAPIAN_VISIBILITY_DEFAULT KMeans {
+    std::vector<Point> docs;
+    std::vector<Centroid> centroids;
+    unsigned int k;
+    std::string mode;
+  public:
+    KMeans(unsigned int k_) : k(k_) { mode = "random"; }
+    KMeans(unsigned int k_, std::string mode_) : k(k_), mode(mode_) {}
+    /** Checks whether the current state of KMeans has converged
+     *  by checking for change in centroid of the clusters
+     */
+    bool converge(std::vector<Centroid> &previous, std::vector<Centroid> &current);
+
+    /** Initialization of centroids using a certain specified method
+     *  Current methods that are supported :
+     *  random - Random Initialization
+     *  kmeanspp - KMeans++ Initialization
+     */
+    void initialize_centroids(ClusterSet &cset);
```

Comment:
Are these four methods really public? They feel internal to the implementation of `cluster()`.